File: instrumentation/java.completable-future-jdk8/src/main/java/util/TokenAwareRunnable.java
package util;

import com.newrelic.agent.bridge.AgentBridge;

import static util.TokenAndRefUtils.clearThreadTokenAndRefCountAndTxn;
import static util.TokenAndRefUtils.getThreadTokenAndRefCount;
import static util.TokenAndRefUtils.logTokenInfo;
import static util.TokenAndRefUtils.setThreadTokenAndRefCount;

public final class TokenAwareRunnable implements Runnable {
    private final Runnable delegate;

    private AgentBridge.TokenAndRefCount tokenAndRefCount;

    public TokenAwareRunnable(Runnable delegate) {
        this.delegate = delegate;
        //get token state from calling Thread
        this.tokenAndRefCount = getThreadTokenAndRefCount();
        logTokenInfo(tokenAndRefCount, "TokenAwareRunnable token info set");
    }

    @Override
    public void run() {
        try {
            if (delegate != null) {
                logTokenInfo(tokenAndRefCount, "Token info set in thread");
                setThreadTokenAndRefCount(tokenAndRefCount);
                delegate.run();
            }
        } finally {
            logTokenInfo(tokenAndRefCount, "Clearing token info from thread ");
            clearThreadTokenAndRefCountAndTxn(tokenAndRefCount);
        }
    }
}

File: instrumentation/cats-effect-2/src/main/scala/com/nr/agent/instrumentation/TokenAwareRunnable.scala
package com.nr.agent.instrumentation

import com.newrelic.agent.bridge.{AgentBridge}
import com.nr.agent.instrumentation.TokenAwareRunnable._

import java.util.concurrent.atomic.AtomicInteger
import java.util.logging.Level

final class TokenAwareRunnable(val delegate: Runnable) extends Runnable {
  private val tokenAndRefCount:AgentBridge.TokenAndRefCount = getThreadTokenAndRefCount
  logTokenInfo(tokenAndRefCount, "EC TokenAwareRunnable token info set")
  clearTxn()

  override def run(): Unit = {
    try if (delegate != null) {
      setThreadTokenAndRefCount(tokenAndRefCount)
      delegate.run()
    }
    finally {
      clearThreadTokenAndRefCountAndTxn(tokenAndRefCount)
    }
  }
}

object TokenAwareRunnable {
  def getThreadTokenAndRefCount: AgentBridge.TokenAndRefCount = {
    val activeToken = AgentBridge.activeToken.get
    val txn = AgentBridge.getAgent.getTransaction(false)
    if(activeToken != null) {
        if(activeToken.refCount == null) {
          activeToken.refCount = new AtomicInteger(1)
        } else {
          activeToken.refCount.incrementAndGet()
        }
        activeToken
    } else if (txn != null) {
      new AgentBridge.TokenAndRefCount(txn.getToken, AgentBridge.getAgent.getTracedMethod, new AtomicInteger(1))
    } else {
      null
    }
  }

  def clearTxn(): Unit = {
    AgentBridge.activeToken.remove()
    val txn = AgentBridge.getAgent.getTransaction(false)
    if(txn != null) {
      txn.clearTransaction()
      logFinest(s"Txn ${txn.hashCode()} cleared from Thread: ${Thread.currentThread().getName}")
    }
  }

  def setThreadTokenAndRefCount(tokenAndRefCount: AgentBridge.TokenAndRefCount): Unit = {
    if (tokenAndRefCount != null && tokenAndRefCount.token != null) {
      AgentBridge.activeToken.set(tokenAndRefCount)
      tokenAndRefCount.token.link
    }
  }

  def clearThreadTokenAndRefCountAndTxn(tokenAndRefCount: AgentBridge.TokenAndRefCount): Unit = {
    AgentBridge.activeToken.remove()
    if (tokenAndRefCount != null && tokenAndRefCount.refCount.decrementAndGet == 0) {
      tokenAndRefCount.token.expire
      tokenAndRefCount.token = null
    }
  }

  def logTokenInfo(tokenAndRefCount: AgentBridge.TokenAndRefCount, msg: String): Unit =
      if (AgentBridge.getAgent.getLogger.isLoggable(Level.FINEST)) {
        logFinest{
          val tokenMsg = if (tokenAndRefCount != null && tokenAndRefCount.token != null) {
            s"[${tokenAndRefCount.token}:${tokenAndRefCount.token.getTransaction}:${tokenAndRefCount.refCount.get}]"
          } else {
            "[Empty token]"
          }

          s"[${Thread.currentThread().getName}] ${tokenMsg}: token info ${msg}"
        }
     }

  def logFinest(msg: => String): Unit =
    if (AgentBridge.getAgent.getLogger.isLoggable(Level.FINEST)) {
      AgentBridge.getAgent.getLogger.log(Level.FINEST, msg)
    }

}
